scipy/stats/_distn_infrastructure.py

```diff
@@ -3836,3 +3884,423 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+        self.dist = dist
+        self.data = data
+        self.discrete = discrete
+        self.pxf = getattr(dist, "pmf", None) or getattr(dist, "pdf", None)
+        self.success = getattr(res, "success", None)
+        self.message = getattr(res, "message", None)
+
+        shape_names = [] if dist.shapes is None else dist.shapes.split(", ")
+        if getattr(dist, "pdf", False):
+            FitShapes = namedtuple('FitShapes', shape_names + ['loc', 'scale'])
+        elif getattr(dist, "pmf", False):
+            FitShapes = namedtuple('FitShapes', shape_names + ['loc'])
+
+        self.params = FitShapes(*res.x)
+        self.objective_val = getattr(res, "fun", None)
+        if self.objective_val is None:
+            self.objective_val = nllf(res.x)
+        self.nllf = nllf
+
+
+    def __repr__(self):
+        keys = ["dist", "data", "nllf", "params", "objective_val",
+                "success", "message"]
+        m = max(map(len, keys)) + 1
+        return '\n'.join([key.rjust(m) + ': ' + repr(getattr(self, key))
+                          for key in keys if getattr(self, key) is not None])
+
+    def plot(self):
+        """Visualize the fit result.
+
+        Superposes the PDF/PMF of the fitted distribution over a normalized
+        histogram of the data.
+
+        Available only if ``matplotlib`` is installed.
+
+        Returns
+        -------
+        fig : matplotlib.figure.Figure
+            A matplotlib Figure object.
+        ax : matplotlib.axes.Axes
+            A matplotlib Axes object.
+        """
+        figure, ax = plt.subplots(nrows=1, ncols=1)
+        fit_params = np.atleast_1d(self.params)
+        support = self.dist.support(*fit_params)
+        lb = support[0] if np.isfinite(support[0]) else min(self.data)
+        ub = support[1] if np.isfinite(support[1]) else max(self.data)
+
+        if self.discrete:
+            x = np.arange(lb, ub + 2)
+            y = self.pxf(x, *fit_params)
+            ax.plot(x[:-1], y[:-1], '--', label='Fit Distribution PMF')
+            options = dict(density=True, bins=x, align='left')
+            ax.xaxis.set_major_locator(MaxNLocator(integer=True))
+
+        else:
+            x = np.linspace(lb, ub, 200)
+            y = self.pxf(x, *fit_params)
+            ax.plot(x, y, '--', label='Fit Distribution PDF')
+            options = dict(density=True, bins=50, align='mid')
+
+        ax.hist(self.data, label="Histogram of Data", **options)
+        ax.set_title(f"{self.dist.name} Fit")
+        ax.legend(*ax.get_legend_handles_labels())
+        plt.show()
+        return figure, ax
+
+
+def fit(dist, data, shape_bounds=None, *, loc_bounds=None, scale_bounds=None,
+        optimizer=optimize.differential_evolution):
+    r"""Fit a discrete or continuous distribution to data
+
+    Given a distribution, data, and bounds for the shape parameters of the
+    distribution, return maximum likelihood estimates of the shape parameters.
```

Comment:
```suggestion
    Given a distribution, data, and bounds for the parameters of the
    distribution, return maximum likelihood estimates of the parameters.
```

Obviously this can be generalized to method of moments, etc., in the future. Keep this in mind elsewhere to make sure we're leaving the door open to that.